lib/ace/plugin_cache.rb

```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+
+require 'fileutils'
+require 'puppet/configurer'
+require 'concurrent'
+require 'ace/fork_util'
+
+module ACE
+  class PluginCache
+    attr_reader :config, :cache_dir_mutex, :uri
+    def initialize(config)
+      @config = config
+      @uri = URI.parse(config['file-server-uri'])
+      @cache_dir_mutex = Concurrent::ReadWriteLock.new
+    end
+
+    def setup
+      FileUtils.mkdir_p(config['cache-dir'])
+      self
+    end
+
+    # @returns the tmp libdir directory which will be where catalogs get executed from
+    def sync(*args)
+      ForkUtil.isolate { sync_core(*args) }
+    end
+
+    # the Puppet[:libdir] will point to a tmp location
+    # where the contents from the pluginsync dest is copied
+    # too.
+    def libdir(plugin_dest)
+      tmpdir = Dir.mktmpdir(['plugins', plugin_dest])
```

Comment:
These `tmpdir`s need to be cleaned up at one point. Since the caller will control the whole lifecycle, providing an appropriate method to do so is sufficient.